pandas/core/dtypes/cast.py

```diff
@@ -1018,6 +1019,78 @@ def soft_convert_objects(
     return values
 
 
+def convert_dtypes(
+    input_array,
+    convert_string: bool = True,
+    convert_integer: bool = True,
+    convert_boolean: bool = True,
+):
+    """
+        Convert objects to best possible type, and optionally,
+        to types supporting ``pd.NA``.
+
+        Parameters
+        ----------
+        input_array : ExtensionArray or PandasArray
+        convert_string : bool, default True
+            Whether object dtypes should be converted to ``StringDtype()``.
+        convert_integer : bool, default True
+            Whether, if possible, conversion can be done to integer extension types.
+        convert_boolean : bool, defaults True
+            Whether object dtypes should be converted to ``BooleanDtypes()``.
+
+        Returns
+        -------
+        Array
+            new dtype
+    """
+
+    if convert_string or convert_integer or convert_boolean:
+        try:
+            inferred_dtype = lib.infer_dtype(input_array)
+        except ValueError:
+            inferred_dtype = input_array.dtype
+        if not convert_string and is_string_dtype(inferred_dtype):
```

Comment:
can you comment on this case here

Reply:
I added a comment, but discussion is here https://github.com/pandas-dev/pandas/pull/30929#discussion_r366465412

The issue is that if you pass in a `Period` to `lib.infer_dtype()` you get a `ValueError`. Reported in #23553

Reply:
in-line comments are *way* more useful than a github comment as they are readable by anyone (you can link to that discussion if you want)

Reply:
I've added an inline comment with a ref to the Github issue.